File: src/core/MapEngine.js
const EventEmitter = require('events');
const CoordinateSystem = require('./CoordinateSystem');
const TileManager = require('./TileManager');
const CacheManager = require('./CacheManager');
const Logger = require('../monitor/Logger');

/**
 * MapEngine - 地图服务引擎核心
 * 负责地图渲染、瓦片管理、坐标转换等核心功能
 */
class MapEngine extends EventEmitter {
  constructor(options = {}) {
    super();
    this.name = 'MapEngine';
    this.version = '1.0.0';
    this.logger = new Logger();
    
    this.options = {
      maxZoom: options.maxZoom || 20,
      minZoom: options.minZoom || 0,
      tileSize: options.tileSize || 256,
      cacheSize: options.cacheSize || 1000,
      projections: options.projections || ['EPSG:4326', 'EPSG:3857']
    };
    
    // 初始化子模块
    this.coordinateSystem = new CoordinateSystem();
    this.tileManager = new TileManager(this.options);
    this.cacheManager = new CacheManager(this.options.cacheSize);
    
    this.layers = new Map();
    this.styles = new Map();
    this.isInitialized = false;
  }
  
  /**
   * 初始化地图引擎
   */
  async initialize() {
    try {
      await this.tileManager.initialize();
      await this.cacheManager.initialize();
      this.isInitialized = true;
      this.logger.info({ message: 'MapEngine initialized' });
      this.emit('initialized');
      return true;
    } catch (error) {
      this.logger.error({ message: 'MapEngine initialization failed', error: error.message });
      throw error;
    }
  }
  
  /**
   * 添加图层
   */
  addLayer(layerId, layerConfig) {
    if (this.layers.has(layerId)) {
      throw new Error(`Layer ${layerId} already exists`);
    }
    
    const layer = {
      id: layerId,
      name: layerConfig.name || layerId,
      type: layerConfig.type || 'raster',
      visible: layerConfig.visible !== false,
      opacity: layerConfig.opacity || 1.0,
      zIndex: layerConfig.zIndex || 0,
      source: layerConfig.source,
      config: layerConfig
    };
    
    this.layers.set(layerId, layer);
    this.logger.info({ message: 'Layer added', layerId });
    this.emit('layer:added', layer);
    return layer;
  }
  
  /**
   * 移除图层
   */
  removeLayer(layerId) {
    if (!this.layers.has(layerId)) {
      throw new Error(`Layer ${layerId} not found`);
    }
    
    this.layers.delete(layerId);
    this.logger.info({ message: 'Layer removed', layerId });
    this.emit('layer:removed', { layerId });
  }
  
  /**
   * 获取瓦片
   */
  async getTile(z, x, y, options = {}) {
    const cacheKey = `tile:${z}:${x}:${y}`;
    
    // 检查缓存
    let tileData = this.cacheManager.get(cacheKey);
    if (tileData) {
      this.emit('tile:cached', { z, x, y });
      return tileData;
    }
    
    try {
      tileData = await this.tileManager.getTile(z, x, y, options);
      this.cacheManager.set(cacheKey, tileData);
      this.emit('tile:loaded', { z, x, y });
      return tileData;
    } catch (error) {
      this.logger.error({ message: 'Failed to get tile', z, x, y, error: error.message });
      throw error;
    }
  }
  
  /**
   * 坐标转换
   */
  transformCoordinates(coords, fromProj, toProj) {
    return this.coordinateSystem.transform(coords, fromProj, toProj);
  }
  
  /**
   * 获取地图信息
   */
  getMapInfo() {
    return {
      name: this.name,
      version: this.version,
      options: this.options,
      layers: Array.from(this.layers.values()),
      cacheStats: this.cacheManager.getStats(),
      isInitialized: this.isInitialized
    };
  }
  
  /**
   * 执行空间查询
   */
  async spatialQuery(bbox, options = {}) {
    const results = [];
    const layers = options.layers || Array.from(this.layers.keys());
    
    for (const layerId of layers) {
      const layer = this.layers.get(layerId);
      if (layer && layer.visible) {
        const layerResults = await this.tileManager.query(layerId, bbox);
        results.push({
          layerId,
          data: layerResults
        });
      }
    }
    
    return results;
  }
  
  /**
   * 清空缓存
   */
  clearCache() {
    this.cacheManager.clear();
    this.logger.info({ message: 'Cache cleared' });
    this.emit('cache:cleared');
  }
  
  /**
   * 获取性能指标
   */
  getMetrics() {
    return {
      cacheStats: this.cacheManager.getStats(),
      tileManagerStats: this.tileManager.getStats(),
      layersCount: this.layers.size
    };
  }
  
  /**
   * 关闭引擎
   */
  async shutdown() {
    await this.tileManager.shutdown();
    await this.cacheManager.shutdown();
    this.isInitialized = false;
    this.logger.info({ message: 'MapEngine shutdown' });
    this.emit('shutdown');
  }
}

module.exports = MapEngine;

File: src/main/java/com/mapengine/core/MapEngine.java
package com.mapengine.core;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class MapEngine {
    private static final Logger logger = LoggerFactory.getLogger(MapEngine.class);

    private final String name = "MapEngine";
    private final String version = "1.0.0";

    private final MapEngineOptions options;
    private final CoordinateSystem coordinateSystem;
    private final TileManager tileManager;
    private final CacheManager cacheManager;
    private final Map<String, Layer> layers = new LinkedHashMap<>();
    private volatile boolean initialized = false;

    public MapEngine() {
        this(new MapEngineOptions());
    }

    public MapEngine(MapEngineOptions options) {
        this.options = options != null ? options : new MapEngineOptions();
        this.coordinateSystem = new CoordinateSystem();
        this.tileManager = new TileManager(this.options);
        this.cacheManager = new CacheManager(this.options.getCacheSize());
    }

    public CompletableFuture<Void> initialize() {
        return CompletableFuture.runAsync(() -> {
            try {
                tileManager.initialize();
                cacheManager.initialize();
                initialized = true;
                logger.info("MapEngine initialized");
            } catch (Exception e) {
                logger.error("MapEngine initialization failed", e);
                throw new RuntimeException(e);
            }
        });
    }

    public synchronized Layer addLayer(String layerId, LayerConfig config) {
        if (layers.containsKey(layerId)) {
            throw new IllegalArgumentException("Layer " + layerId + " already exists");
        }
        Layer layer = new Layer(layerId, config);
        layers.put(layerId, layer);
        logger.info("Layer added: {}", layerId);
        return layer;
    }

    public synchronized void removeLayer(String layerId) {
        if (!layers.containsKey(layerId)) throw new NoSuchElementException("Layer not found: " + layerId);
        layers.remove(layerId);
        logger.info("Layer removed: {}", layerId);
    }

    public TileData getTile(int z, int x, int y) throws Exception {
        String cacheKey = String.format("tile:%d:%d:%d", z, x, y);
        TileData cached = cacheManager.get(cacheKey);
        if (cached != null) {
            return cached;
        }
        TileData tile = tileManager.getTile(z, x, y);
        cacheManager.set(cacheKey, tile);
        return tile;
    }

    public double[] transformCoordinates(double[] coords, String fromProj, String toProj) {
        return coordinateSystem.transform(coords, fromProj, toProj);
    }

    public MapInfo getMapInfo() {
        return new MapInfo(name, version, options, new ArrayList<>(layers.values()), cacheManager.getStats(), initialized);
    }

    public void clearCache() {
        cacheManager.clear();
        logger.info("Cache cleared");
    }

    public Metrics getMetrics() {
        return new Metrics(cacheManager.getStats(), tileManager.getStats(), layers.size());
    }

    public void shutdown() {
        try {
            tileManager.shutdown();
        } catch (Exception e) {
            logger.warn("Error shutting down tileManager", e);
        }
        try {
            cacheManager.shutdown();
        } catch (Exception e) {
            logger.warn("Error shutting down cacheManager", e);
        }
        initialized = false;
        logger.info("MapEngine shutdown");
    }

    // small DTOs
    @Getter
    @RequiredArgsConstructor
    public static class Metrics {
        private final Object cacheStats;
        private final Object tileManagerStats;
        private final int layersCount;
    }
}
